homeassistant/components/zone/trigger.py

```diff
@@ -67,16 +64,16 @@ async def async_attach_trigger(
     """Listen for state changes based on configuration."""
     trigger_data = trigger_info["trigger_data"]
     entity_id: list[str] = config[CONF_ENTITY_ID]
-    zone_entity_id = config.get(CONF_ZONE)
-    event = config.get(CONF_EVENT)
+    zone_entity_id = config[CONF_ZONE]
+    event = config[CONF_EVENT]
     job = HassJob(action)
 
     @callback
-    def zone_automation_listener(zone_event):
+    def zone_automation_listener(zone_event: Event) -> None:
         """Listen for state changes and calls action."""
         entity = zone_event.data.get("entity_id")
         from_s = zone_event.data.get("old_state")
-        to_s = zone_event.data.get("new_state")
+        to_s = zone_event.data["new_state"]
```

Comment:
I think that `old_state` might be None, but surely `new_state` should always be present.
```suggestion
        to_s = zone_event.data["new_state"]
```

Reply:
> I think that `old_state` might be None, but surely `new_state` should always be present.

Not so sure about that. All the code I've seen usually treats both (new and old state) as optional.
https://github.com/home-assistant/core/blob/5c8e8e48607f9e3ab13a17f6566fd50d09c448c7/homeassistant/core.py#L1370-L1375

Even this method does later check for `None`
https://github.com/home-assistant/core/blob/5c8e8e48607f9e3ab13a17f6566fd50d09c448c7/homeassistant/components/zone/trigger.py#L97

Maybe a workaround could be to adjust the conditional here to also check for `to_s` in the `or` clause.
https://github.com/home-assistant/core/blob/5c8e8e48607f9e3ab13a17f6566fd50d09c448c7/homeassistant/components/zone/trigger.py#L81-L85

Reply:
The main issue is when calling `to_s.context` in `hass.async_run_hass_job`
I've added an assert, but I'm not convinced it's the best way to handle it.

Reply:
> The main issue is when calling `to_s.context` in `hass.async_run_hass_job`
> I've added an assert, but I'm not convinced it's the best way to handle it.

The context for `TriggerActionType` is optional. Why not do `to_s.context if to_s else None`?